lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -276,536 +344,505 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
-      m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++,
-                                                          Util::Abi::UserDataMapping::ViewId);
-    }
 
-    if (reserveEsGsLdsSize) {
-      *inRegMask |= 1ull << argTys.size();
-      argTys.push_back(Type::getInt32Ty(*m_context));
-      m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++,
-                                                          Util::Abi::UserDataMapping::EsGsLdsSize);
-    }
-    break;
+  // Remove trailing padding nodes (compute shader).
+  while (!unspilledArgs.empty() && unspilledArgs.back().isPadding) {
+    userDataIdx -= unspilledArgs.back().argDwordSize;
+    unspilledArgs.pop_back();
   }
-  case ShaderStageGeometry: {
-    // NOTE: The user data to emulate gl_ViewIndex is somewhat common. To make it consistent for GFX9
-    // merged shader, we place it prior to any other special user data.
-    if (enableMultiView) {
-      *inRegMask |= 1ull << argTys.size();
-      entryArgIdxs.gs.viewIndex = argTys.size();
-      argTys.push_back(Type::getInt32Ty(*m_context)); // View Index
-      m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++,
-                                                          Util::Abi::UserDataMapping::ViewId);
-      ++userDataIdx;
-    }
 
-    if (reserveEsGsLdsSize) {
-      *inRegMask |= 1ull << argTys.size();
-      argTys.push_back(Type::getInt32Ty(*m_context));
-      m_pipelineState->getPalMetadata()->setUserDataEntry(m_shaderStage, userDataIdx++,
-                                                          Util::Abi::UserDataMapping::EsGsLdsSize);
-    }
-    break;
-  }
-  case ShaderStageCompute: {
-    // Emulate gl_NumWorkGroups via user data registers
-    if (builtInUsage.cs.numWorkgroups) {
-      // NOTE: Pointer must be placed in even index according to LLVM backend compiler.
-      if ((userDataIdx % 2) != 0) {
-        *inRegMask |= 1ull << argTys.size();
-        argTys.push_back(Type::getInt32Ty(*m_context));
-        userDataIdx += 1;
+  // Add the spill table pointer and the extra args to unspilledArgs. How that is done is different between
+  // a compute shader and a graphics shader.
+  if (m_shaderStage == ShaderStageCompute) {
+    // For compute shader:
+    // If we need the spill table pointer, it must go into s11. Insert padding if necessary.
+    if (!spillTableArg.empty()) {
+      if (userDataIdx != userDataEnd) {
+        assert(userDataIdx <= userDataEnd);
+        unspilledArgs.push_back(UserDataArg(VectorType::get(builder.getInt32Ty(), userDataEnd - userDataIdx),
+                                            Util::Abi::UserDataMapping::Invalid, nullptr, /*padding=*/true));
+      }
+      unspilledArgs.push_back(spillTableArg.front());
+      // That then leaves us at a 2-aligned reg, ready for any extraUserDataArgs.
+    } else if (!extraUserDataArgs.empty()) {
```

Comment:
Why else-if? I'd expect situations like:
```
s12 = spill table
s13 = <padding>
s14 = workgroup_ptr_lo
s15 = workgroup_ptr_hi
```

Perhaps keep this if separate, and move the insertion of extraUserDataArgs inside it to be clearer.

Reply:
Yes, I think this logic is wrong. It was ok when I thought that the spill table has to be in s11. :-)